modules/gapi/src/executor/gasync.cpp

```diff
@@ -0,0 +1,198 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "opencv2/gapi/gcomputation_async.hpp"
+#include "opencv2/gapi/gcomputation.hpp"
+#include "opencv2/gapi/gcompiled_async.hpp"
+#include "opencv2/gapi/gcompiled.hpp"
+
+#include <condition_variable>
+
+#include <future>
+#include <condition_variable>
+//#include <chrono>
+#include <stdexcept>
+#include <queue>
+
+namespace {
+    //This is a tool to move initialize captures of a lambda in C++11
+    template<typename T>
+    struct move_through_copy{
+       T value;
+       move_through_copy(T&& g) : value(std::move(g)) {}
+       move_through_copy(move_through_copy&&) = default;
+       move_through_copy(move_through_copy const& lhs) : move_through_copy(std::move(const_cast<move_through_copy&>(lhs))) {}
+    };
+}
+
+namespace cv {
+namespace gapi {
+namespace wip  {
+
+namespace impl{
+
+class async_service {
+
+    std::mutex mtx;
+    std::condition_variable cv;
+    std::queue<std::function<void()>> q;
+    std::atomic<bool> exiting           = {false};
+    std::atomic<bool> thread_started    = {false};
+
+    std::thread thrd;
+
+public:
+    async_service() = default ;
+
+    void add_task(std::function<void()>&& t){
+        if (!thread_started)
+        {
+            //thread has not been started yet, so start it
+            //try to Compare And Swap the flag, false -> true
+            //If there are multiple threads - only single one will succeed in changing the value.
+            bool expected = false;
+            if (thread_started.compare_exchange_strong(expected, true))
+            {
+                //have won (probable) race - so actually start the thread
+                thrd = std::thread {[this](){
+                    //move the whole queue into local instance in order to minimize time the protecting lock is held
+                    decltype(q) second_q;
+                    while (!exiting){
+                        std::unique_lock<std::mutex> lck{mtx};
+                        if (q.empty())
+                        {
+                            //block current thread until arrival of exit request or new elements
+                            cv.wait(lck, [&](){ return exiting || !q.empty();});
+                        }
+                        //usually swap for std::queue is plain pointers exchange, so relatively cheap
+                        q.swap(second_q);
+                        lck.unlock();
+
+                        while (!second_q.empty())
```

Comment:
should this `while` condition be extended with `exiting` check as well?

What is the policy on `exiting`? Should be wait for all issues request or just "forget" it?